File: Algorithms/src/729. My Calendar I/Code/solution.py
class MyCalendar:
    
    def __init__(self):
        # Initialize an empty list to store the booking intervals
        # Each booking will be a tuple (start, end) representing the start and end time of the event
        self.bookings = []

    def book(self, start: int, end: int) -> bool:
        """
        Attempt to book a new event. The event is represented by its start and end time.
        If the event does not overlap with any existing bookings, it is added to the list.
        If it overlaps, the booking is rejected.
        
        :param start: Start time of the event (inclusive)
        :param end: End time of the event (exclusive)
        :return: True if the event is successfully booked, False otherwise.
        """
        
        # Iterate through all the existing bookings to check for any overlap
        for existingStart, existingEnd in self.bookings:
            # Check if the new event overlaps with the existing one
            # Two intervals [start1, end1) and [start2, end2) overlap if:
            # - the start of one event is before the end of the other
            # - AND the end of one event is after the start of the other
            # This condition is a simplified overlap check:
            #   If (start < existingEnd and end > existingStart), it means the events overlap.
            if start < existingEnd and end > existingStart:
                # If the new event overlaps with an existing one, reject the booking
                return False
        
        # If no overlap is found, add the new event to the list of bookings
        # Store the event as a tuple (start, end)
        self.bookings.append((start, end))
        
        # Return True to indicate that the booking was successful
        return True


# Example usage:
# Create an instance of MyCalendar
# obj = MyCalendar()
# Try booking a new event with start and end times
# param_1 = obj.book(start, end)

File: Algorithms/src/729. My Calendar I/Code/solution.js
// Define the MyCalendar class constructor
var MyCalendar = function () {
  // Initialize an empty array to store all the booked events
  this.bookings = [];
};

/**
 * Function to book an event in the calendar
 * @param {number} start - The start time of the event
 * @param {number} end - The end time of the event
 * @return {boolean} - Returns true if the event can be booked, false if there is a conflict
 */
MyCalendar.prototype.book = function (start, end) {
  // Loop through all previously booked events to check for any overlaps
  for (let event of this.bookings) {
    let [existingStart, existingEnd] = event; // Destructure to get the start and end times of the booked event

    // Check if the new event overlaps with the current booked event
    // Overlap occurs if the new event's start time is before an existing event's end time
    // AND the new event's end time is after an existing event's start time
    if (start < existingEnd && end > existingStart) {
      // If overlap is detected, return false to indicate the event can't be booked
      return false;
    }
  }

  // If no overlap is found, add the new event (start, end) to the bookings list
  this.bookings.push([start, end]);

  // Return true to indicate the event was successfully booked
  return true;
};

/**
 * Example usage:
 * var obj = new MyCalendar();
 * var param_1 = obj.book(start, end); // Call the book method with event start and end times
 */
